// @bsmnt/scrollytelling/dist/debugger-Z3ISUZF3.mjs
import { internalEventEmmiter, clsx, __spreadProps, __spreadValues } from './chunk-WGDYDWJO.mjs';
import { useState, useEffect, useRef } from 'react';
import { gsap } from 'gsap';
import { jsx, Fragment, jsxs } from 'react/jsx-runtime';
import { Portal } from '@radix-ui/react-portal';

// esbuild-scss-modules-plugin:./visualizer.module.scss
var digest = "fec6b8638e76466d1b7c01af48179311896075973b211358ed0723298a2a19d4";
var classes = { "root": "_root_hgyrj_2", "initialized": "_initialized_hgyrj_29", "header": "_header_hgyrj_32", "actions": "_actions_hgyrj_49", "button": "_button_hgyrj_54", "scrollToRoot": "_scrollToRoot_hgyrj_68", "text": "_text_hgyrj_75", "main": "_main_hgyrj_78", "timeline": "_timeline_hgyrj_87", "guides": "_guides_hgyrj_93", "guide": "_guide_hgyrj_93", "percent": "_percent_hgyrj_109", "progress": "_progress_hgyrj_118", "trail": "_trail_hgyrj_123", "gradient": "_gradient_hgyrj_132", "marker": "_marker_hgyrj_141", "thumb": "_thumb_hgyrj_151", "tweens": "_tweens_hgyrj_165", "row": "_row_hgyrj_171", "footer": "_footer_hgyrj_180", "version": "_version_hgyrj_192", "bsmnt": "_bsmnt_hgyrj_195", "tween": "_tween_hgyrj_165", "waypoint": "_waypoint_hgyrj_224", "onCall": "_onCall_hgyrj_238", "onReverseCall": "_onReverseCall_hgyrj_239", "active": "_active_hgyrj_250", "selectWrapper": "_selectWrapper_hgyrj_264", "select": "_select_hgyrj_264", "arrow": "_arrow_hgyrj_280", "highlight": "_highlight_hgyrj_292" };
var css = `@import url("https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap");
._root_hgyrj_2 {
  --color-gray-lighter: #a3a3a3;
  --color-gray-light: #383838;
  --color-gray: #232323;
  --color-gray-dark: #121212;
  --color-black: #000000;
  --color-white: #ffffff;
  --color-orange: #ff4d00;
  display: flex;
  flex-direction: column;
  width: 90vw;
  max-width: 800px;
  height: auto;
  max-height: 50vh;
  position: fixed;
  z-index: 9999;
  background: #2b2d37;
  border: 1px solid var(--color-gray-light);
  border-radius: 8px;
  overflow: hidden;
  line-height: 1;
  top: 16px;
  left: 16px;
  font-family: "JetBrains Mono", monospace;
  visibility: hidden;
  color: var(--color-white);
}
._root_hgyrj_2._initialized_hgyrj_29 {
  visibility: visible;
}
._root_hgyrj_2 ._header_hgyrj_32 {
  background: var(--color-gray-dark);
  text-align: center;
  padding: 4px 16px;
  min-height: 44px;
  font-size: 12px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  text-transform: uppercase;
}
._root_hgyrj_2 ._header_hgyrj_32:hover {
  cursor: grab;
}
._root_hgyrj_2 ._header_hgyrj_32:active {
  cursor: grabbing;
}
._root_hgyrj_2 ._header_hgyrj_32 ._actions_hgyrj_49 {
  display: flex;
  gap: 0 8px;
  align-items: center;
}
._root_hgyrj_2 ._header_hgyrj_32 ._actions_hgyrj_49 ._button_hgyrj_54 {
  color: var(--color-white);
  display: flex;
  align-items: center;
  height: 24px;
  padding: 0px 4px;
  border-radius: 4px;
  background: var(--color-gray);
  border: 1px solid var(--color-gray-light);
}
._root_hgyrj_2 ._header_hgyrj_32 ._actions_hgyrj_49 ._button_hgyrj_54 svg {
  width: 14px;
  height: 14px;
}
._root_hgyrj_2 ._header_hgyrj_32 ._actions_hgyrj_49 ._scrollToRoot_hgyrj_68 {
  display: flex;
  align-items: center;
  padding-left: 8px;
  padding-right: 8px;
  gap: 0 8px;
}
._root_hgyrj_2 ._header_hgyrj_32 ._actions_hgyrj_49 ._scrollToRoot_hgyrj_68 ._text_hgyrj_75 {
  transform: translateY(0.05em);
}
._root_hgyrj_2 ._main_hgyrj_78 {
  border-top: 1px solid var(--color-gray-light);
  border-bottom: 1px solid var(--color-gray-light);
  background: var(--color-black);
  padding: 24px 32px 24px;
  flex: 1;
  overflow: auto;
  max-height: 100%;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 {
  margin-top: 32px;
  position: relative;
  height: 100%;
  width: 100%;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._guides_hgyrj_93 {
  position: absolute;
  height: 100%;
  bottom: 0;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._guides_hgyrj_93 ._guide_hgyrj_93 {
  position: absolute;
  height: 100%;
  width: 1px;
  border-right: 1px dashed rgba(255, 255, 255, 0.5333333333);
  font-size: 14px;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._guides_hgyrj_93 ._guide_hgyrj_93 ._percent_hgyrj_109 {
  padding-bottom: 16px;
  position: absolute;
  top: 0;
  left: 50%;
  line-height: 1;
  transform: translateY(-100%) translateX(-50%);
  color: rgba(255, 255, 255, 0.5333333333);
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._progress_hgyrj_118 {
  pointer-events: none;
  position: absolute;
  inset: 0;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._progress_hgyrj_118 ._trail_hgyrj_123 {
  position: absolute;
  height: 100%;
  top: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  overflow: hidden;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._progress_hgyrj_118 ._trail_hgyrj_123 ._gradient_hgyrj_132 {
  position: absolute;
  height: 100%;
  top: 0;
  bottom: 0;
  width: 200px;
  background: linear-gradient(270deg, rgba(255, 255, 255, 0.14) 0.53%, rgba(255, 255, 255, 0) 100%);
  transform: translateX(-100%);
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._progress_hgyrj_118 ._marker_hgyrj_141 {
  position: absolute;
  height: 100%;
  top: 0;
  bottom: 0;
  width: 1px;
  background: var(--color-white);
  box-sizing: content-box;
  font-size: 14px;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._progress_hgyrj_118 ._marker_hgyrj_141 ._thumb_hgyrj_151 {
  top: 0;
  transform: translateX(-50%) translateY(-100%);
  left: 50%;
  position: absolute;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-bottom: 2px;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._progress_hgyrj_118 ._marker_hgyrj_141 ._thumb_hgyrj_151 ._percent_hgyrj_109 {
  line-height: 1;
  padding-bottom: 3px;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._tweens_hgyrj_165 {
  position: relative;
  height: 100%;
  display: flex;
  flex-direction: column;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._tweens_hgyrj_165 ._row_hgyrj_171 {
  padding: 4px 0;
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._tweens_hgyrj_165 ._row_hgyrj_171:nth-child(odd) {
  background: rgba(20, 21, 26, 0.3764705882);
}
._root_hgyrj_2 ._main_hgyrj_78 ._timeline_hgyrj_87 ._tweens_hgyrj_165 ._row_hgyrj_171:nth-child(even) {
  background: rgba(20, 21, 26, 0.5647058824);
}
._root_hgyrj_2 ._footer_hgyrj_180 {
  background: var(--color-gray-dark);
  text-transform: uppercase;
  text-align: center;
  padding: 4px 16px;
  min-height: 44px;
  font-size: 14px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  color: var(--color-white);
}
._root_hgyrj_2 ._footer_hgyrj_180 ._version_hgyrj_192 {
  color: var(--color-gray-lighter);
}
._root_hgyrj_2 ._footer_hgyrj_180 ._bsmnt_hgyrj_195 {
  color: var(--color-orange);
}

._tween_hgyrj_165 {
  --duration-percentage: 0;
  --start-offset-percentage: 0;
  left: var(--start-offset-percentage);
  /* (1px for outline width + 1px for outline offset) */
  width: calc(var(--duration-percentage) - 3px);
  cursor: default;
  font-size: 12px;
  color: var(--color-white);
  position: relative;
  display: flex;
  align-items: center;
  min-height: 12px;
  height: 18px;
  background: white;
  padding-left: 8px;
  font-weight: 500;
  border-radius: 0px 4px 4px 0px;
  margin: 0 2px;
  outline: 1px solid var(--color-gray-light);
  outline-offset: 1px;
  overflow: hidden;
  white-space: nowrap;
}

._waypoint_hgyrj_224 {
  --start-offset-percentage: 0;
  position: relative;
  left: var(--start-offset-percentage);
  width: 22px;
  height: 22px;
  display: flex;
  align-items: center;
  justify-content: center;
  background: var(--color-gray);
  border: 1px solid var(--color-gray-light);
  border-radius: 6px;
  transform: translateX(-50%);
}
._waypoint_hgyrj_224 ._onCall_hgyrj_238,
._waypoint_hgyrj_224 ._onReverseCall_hgyrj_239 {
  opacity: 0;
  transition-property: transform opacity;
  transition-duration: 0.2s;
  transition-timing-function: ease-out;
}
._waypoint_hgyrj_224 ._onCall_hgyrj_238 {
  position: absolute;
  right: -8px;
  transform: translateX(50%);
}
._waypoint_hgyrj_224 ._onCall_hgyrj_238._active_hgyrj_250 {
  opacity: 1;
  transform: translateX(100%) scale(1.2);
}
._waypoint_hgyrj_224 ._onReverseCall_hgyrj_239 {
  position: absolute;
  left: -8px;
  transform: translateX(-50%);
}
._waypoint_hgyrj_224 ._onReverseCall_hgyrj_239._active_hgyrj_250 {
  opacity: 1;
  transform: translateX(-100%) scale(1.2);
}

._selectWrapper_hgyrj_264 {
  position: relative;
}
._selectWrapper_hgyrj_264 ._select_hgyrj_264 {
  appearance: none;
  cursor: pointer;
  background: var(--color-gray);
  color: var(--color-white);
  border: 1px solid var(--color-gray-light);
  height: 24px;
  border-radius: 4px;
  width: max-content;
  min-width: 110px;
  padding: 0 8px;
  padding-right: 24px;
}
._selectWrapper_hgyrj_264 ._arrow_hgyrj_280 {
  pointer-events: none;
  position: absolute;
  content: "";
  right: 8px;
  top: 50%;
  width: 12px;
  height: 12px;
  z-index: 10;
  transform: translateY(-50%);
}

._highlight_hgyrj_292 {
  background: rgba(236, 91, 41, 0.6666666667);
  z-index: 999;
}

@media (max-width: 768px) {
  ._root_hgyrj_2 {
    left: 50%;
    transform: translateX(-50%);
  }
  ._root_hgyrj_2 ._version_hgyrj_192 {
    display: none;
  }
  ._root_hgyrj_2 ._scrollToRoot_hgyrj_68 ._text_hgyrj_75 {
    display: none;
  }
}`;
(function() {
  if (typeof document !== "undefined" && !document.getElementById(digest)) {
    var ele = document.createElement("style");
    ele.id = digest;
    ele.textContent = css;
    document.head.appendChild(ele);
  }
})();
var visualizer_module_default = classes;

// package.json
var package_default = {
  name: "@bsmnt/scrollytelling",
  author: "JB <jb@basement.studio>",
  version: "0.3.3",
  main: "./dist/index.js",
  module: "./dist/index.mjs",
  types: "./dist/index.d.ts",
  source: "./src/index.ts",
  sideEffects: false,
  scripts: {
    dev: "yarn build --watch",
    build: "tsup --config tsup.config.ts"
  },
  peerDependencies: {
    gsap: ">=3",
    react: ">=18",
    "react-dom": ">=18"
  },
  dependencies: {
    "@radix-ui/react-portal": ">=1",
    "@radix-ui/react-slot": ">=1"
  },
  devDependencies: {
    "@types/react": "^18.0.17",
    "@types/react-dom": "^18.0.6",
    "esbuild-scss-modules-plugin": "^1.1.1",
    gsap: "^3.11.5",
    react: "^18.2.0",
    "react-dom": "^18.2.0",
    tsconfig: "*",
    tsup: "^6.2.3",
    "type-fest": "^3.0.0",
    typescript: "^4.8.4"
  }
};

// src/components/debugger/visualizer/helpers.tsx
var highlight = (target) => {
  const highlight2 = document.createElement("div");
  highlight2.style.position = "fixed";
  const bounds = target.getBoundingClientRect();
  highlight2.style.top = `${bounds.top}px`;
  highlight2.style.left = `${bounds.left}px`;
  highlight2.style.width = `${bounds.width}px`;
  highlight2.style.height = `${bounds.height}px`;
  highlight2.classList.add(visualizer_module_default["highlight"]);
  document.body.appendChild(highlight2);
  return () => {
    document.body.removeChild(highlight2);
  };
};
var colors = [
  ["#F87171", "#991B1B"],
  ["#FACC15", "#854D0E"],
  ["#4ADE80", "#166534"],
  ["#2DD4BF", "#115E59"],
  ["#38BDF8", "#075985"],
  ["#818CF8", "#3730A3"],
  ["#C084FC", "#6B21A8"],
  ["#E879F9", "#86198F"],
  ["rgba(244, 114, 182, 0.40)", "#9D174D"]
];
var Tween = ({
  tween,
  root,
  idx
}) => {
  var _a, _b;
  const [isHovering, setIsHovering] = useState(false);
  const [active, setActive] = useState(false);
  useEffect(() => {
    if (isHovering) {
      const cleanups = tween.targets().map((t) => {
        if (t instanceof SVGElement || t instanceof HTMLElement) {
          return highlight(t);
        }
        return void 0;
      });
      return () => {
        cleanups.forEach((clean) => clean && clean());
      };
    }
  }, [isHovering, tween]);
  useEffect(() => {
    const handleUpdate = () => {
      var _a2;
      const progress = (_a2 = root.tween) == null ? void 0 : _a2.progress();
      if (!progress)
        return;
      const start = tween._start / 100;
      const end = (tween._start + tween._dur) / 100;
      if (progress >= start && progress <= end) {
        setActive(true);
      } else {
        setActive(false);
      }
    };
    return internalEventEmmiter.on("timeline:update", handleUpdate);
  }, [root.tween, tween._dur, tween._start]);
  const targetString = tween.targets().map((t) => {
    if (t instanceof SVGElement || t instanceof HTMLElement) {
      return `${t.tagName.toLocaleLowerCase()}${t.id ? `#${t.id}` : ""}${t.classList.length ? "." + t.classList[0] : ""}`;
    }
    if (t instanceof Object) {
      const allKeys = Object.keys(t).filter((k) => k != "_gsap");
      const displayKeys = allKeys.slice(0, 3);
      if (allKeys.length > displayKeys.length) {
        displayKeys.push("...");
      }
      return `${t.constructor.name} { ${displayKeys.join(", ")} }`;
    }
  }).join(", ");
  return /* @__PURE__ */ jsx(
    "div",
    {
      title: targetString,
      className: `${visualizer_module_default["tween"]}${active ? ` ${[visualizer_module_default["active"]]}` : ""}`,
      style: {
        // @ts-ignore
        "--duration-percentage": tween._dur + "%",
        "--start-offset-percentage": tween._start + "%",
        background: "linear-gradient(90deg, transparent 0%, " + ((_a = colors[idx % colors.length]) == null ? void 0 : _a[0]) + " 100%)",
        outlineColor: (_b = colors[idx % colors.length]) == null ? void 0 : _b[1],
        minWidth: 16
      },
      onMouseEnter: () => setIsHovering(true),
      onMouseLeave: () => setIsHovering(false),
      onClick: () => {
        var _a2;
        const st = (_a2 = root.tween) == null ? void 0 : _a2.scrollTrigger;
        if (!st)
          return;
        const start = st.start;
        const end = st.end;
        const scroll = start + (end - start) * (tween._start / 100);
        console.log({ start, end, tweenStart: tween._start });
        window.scrollTo({ top: scroll, behavior: "smooth" });
      },
      children: targetString
    }
  );
};
var Waypoint = ({
  tween,
  root
}) => {
  const [lastState, setLastState] = useState(void 0);
  useEffect(() => {
    if (tween.data.type === "waypoint") {
      tween.data._internalOnCall = () => {
        setLastState("complete");
      };
      tween.data._internalOnReverseCall = () => {
        setLastState("reverse-complete");
      };
    }
  }, [tween.data]);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      style: {
        // @ts-ignore
        "--start-offset-percentage": tween._start + "%"
      },
      className: visualizer_module_default["waypoint"],
      onClick: () => {
        var _a, _b;
        const st = (_a = root.tween) == null ? void 0 : _a.scrollTrigger;
        if (!st)
          return;
        if (tween.data.type === "waypoint") {
          const foundLabel = (_b = root.tween) == null ? void 0 : _b.labels[tween.data.label];
          if (foundLabel) {
            const targetPx = st.labelToScroll(tween.data.label);
            window.scrollTo({ top: targetPx + 0, behavior: "smooth" });
          }
        }
      },
      children: [
        /* @__PURE__ */ jsx(
          "span",
          {
            className: clsx(
              visualizer_module_default["onReverseCall"],
              lastState === "reverse-complete" && visualizer_module_default["active"]
            ),
            children: /* @__PURE__ */ jsx(
              "svg",
              {
                width: "12",
                height: "12",
                viewBox: "0 0 12 12",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
                children: /* @__PURE__ */ jsx(
                  "path",
                  {
                    d: "M6.375 9.75L2.625 6L6.375 2.25M9.375 9.75L5.625 6L9.375 2.25",
                    stroke: "white",
                    strokeLinecap: "round",
                    strokeLinejoin: "round"
                  }
                )
              }
            )
          }
        ),
        /* @__PURE__ */ jsxs(
          "svg",
          {
            width: "12",
            height: "12",
            viewBox: "0 0 12 12",
            fill: "none",
            xmlns: "http://www.w3.org/2000/svg",
            children: [
              /* @__PURE__ */ jsx(
                "path",
                {
                  d: "M1.5 1.5V2.25V1.5ZM1.5 10.5V7.5V10.5ZM1.5 7.5L2.885 7.1535C3.9272 6.89302 5.0282 7.01398 5.989 7.4945L6.043 7.5215C6.98459 7.9921 8.06137 8.11772 9.086 7.8765L10.643 7.5105C10.4523 5.76591 10.4515 4.00577 10.6405 2.261L9.0855 2.627C8.06097 2.86794 6.98439 2.74215 6.043 2.2715L5.989 2.2445C5.0282 1.76398 3.9272 1.64302 2.885 1.9035L1.5 2.25M1.5 7.5V2.25V7.5Z",
                  fill: "white",
                  fillOpacity: "0.12"
                }
              ),
              /* @__PURE__ */ jsx(
                "path",
                {
                  d: "M1.5 1.5V2.25M1.5 2.25L2.885 1.9035C3.9272 1.64302 5.0282 1.76398 5.989 2.2445L6.043 2.2715C6.98439 2.74215 8.06097 2.86794 9.0855 2.627L10.6405 2.261C10.4515 4.00577 10.4523 5.76591 10.643 7.5105L9.086 7.8765C8.06137 8.11772 6.98459 7.9921 6.043 7.5215L5.989 7.4945C5.0282 7.01398 3.9272 6.89302 2.885 7.1535L1.5 7.5M1.5 2.25V7.5M1.5 10.5V7.5",
                  stroke: "white",
                  strokeLinecap: "round",
                  strokeLinejoin: "round"
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsx(
          "span",
          {
            className: clsx(visualizer_module_default["onCall"], lastState === "complete" && visualizer_module_default["active"]),
            children: /* @__PURE__ */ jsx(
              "svg",
              {
                width: "12",
                height: "12",
                viewBox: "0 0 12 12",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
                children: /* @__PURE__ */ jsx(
                  "path",
                  {
                    d: "M5.625 2.25L9.375 6L5.625 9.75M2.625 2.25L6.375 6L2.625 9.75",
                    stroke: "white",
                    strokeLinecap: "round",
                    strokeLinejoin: "round"
                  }
                )
              }
            )
          }
        )
      ]
    }
  );
};
var ProgressStatus = ({ root }) => {
  const [progress, setProgress] = useState(0);
  useEffect(() => {
    const handleUpdate = () => {
      var _a;
      const progress2 = (_a = root == null ? void 0 : root.tween) == null ? void 0 : _a.progress();
      setProgress(progress2 != null ? progress2 : 0);
    };
    return internalEventEmmiter.on("timeline:update", handleUpdate);
  }, [root == null ? void 0 : root.tween]);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    (progress * 100).toFixed(0),
    "%"
  ] });
};
var Select = (props) => {
  return /* @__PURE__ */ jsxs("div", { className: visualizer_module_default["selectWrapper"], children: [
    /* @__PURE__ */ jsx("select", __spreadProps(__spreadValues({}, props), { className: clsx(visualizer_module_default["select"], props.className) })),
    /* @__PURE__ */ jsx(
      "svg",
      {
        className: visualizer_module_default["arrow"],
        width: "12",
        height: "12",
        viewBox: "0 0 12 12",
        fill: "none",
        xmlns: "http://www.w3.org/2000/svg",
        children: /* @__PURE__ */ jsx(
          "path",
          {
            d: "M9.75 4.125L6 7.875L2.25 4.125",
            stroke: "white",
            strokeLinecap: "round",
            strokeLinejoin: "round"
          }
        )
      }
    )
  ] });
};
var Visualizer = () => {
  var _a;
  const panelRef = useRef(null);
  const panelHeaderRef = useRef(null);
  const markerRef = useRef(null);
  const trailRef = useRef(null);
  const [roots, setRoots] = useState([]);
  const [selectedRoot, setSelectedRoot] = useState();
  const [dismiss, setDismiss] = useState(false);
  const [minimize, setMinimize] = useState(false);
  const [initialized, setInitialized] = useState(false);
  const [scrollTop, setScrollTop] = useState();
  const [isUserScroll, setIsUserScroll] = useState(true);
  const root = (_a = roots.find((r) => r.id === selectedRoot)) != null ? _a : roots[0];
  useEffect(() => {
    if (!isUserScroll)
      return;
    const onScroll = () => {
      setScrollTop(window.scrollY);
    };
    window.addEventListener("scroll", onScroll);
    const activeRoot = roots.find(
      (r) => {
        var _a2;
        const progress = (_a2 = r == null ? void 0 : r.tween) == null ? void 0 : _a2.progress();
        const roundedProgress = Math.round(progress * 100) / 100;
        return roundedProgress !== void 0 && (roundedProgress > 0 && roundedProgress < 1);
      }
    );
    if (!activeRoot)
      return;
    setSelectedRoot(activeRoot.id);
    return () => {
      window.removeEventListener("scroll", onScroll);
    };
  }, [isUserScroll, roots, scrollTop]);
  useEffect(() => {
    const handleUpdate = () => {
      var _a2;
      const progress = (_a2 = root == null ? void 0 : root.tween) == null ? void 0 : _a2.progress();
      if (!markerRef.current || !trailRef.current || progress === void 0) {
        return;
      }
      markerRef.current.style.left = `${progress * 100}%`;
      trailRef.current.style.left = `${progress * 100}%`;
    };
    return internalEventEmmiter.on("timeline:update", handleUpdate);
  }, [root == null ? void 0 : root.tween]);
  useEffect(() => {
    const panel = panelRef.current;
    const panelHeader = panelHeaderRef.current;
    const lastPosition = sessionStorage.getItem(
      "@bmsmnt/scrollytelling-visualizer:position"
    );
    if (!panel || !panelHeader)
      return;
    let pos1 = 0, pos2 = 0, pos3 = 0, pos4 = 0;
    if (lastPosition) {
      const { top, left } = JSON.parse(lastPosition);
      panel.style.top = top;
      panel.style.left = left;
    }
    const dragMouseDown = (e) => {
      e = e || window.event;
      const target = e.target;
      if (target instanceof HTMLElement) {
        const hasInteractableClosest = target.closest("button") || target.closest("select") || target.closest("input");
        if (hasInteractableClosest)
          return;
      }
      e.preventDefault();
      pos3 = e.clientX;
      pos4 = e.clientY;
      document.addEventListener("mouseup", closeDragElement);
      document.addEventListener("mousemove", elementDrag);
    };
    const elementDrag = (e) => {
      if (!panel)
        return;
      e = e || window.event;
      e.preventDefault();
      pos1 = pos3 - e.clientX;
      pos2 = pos4 - e.clientY;
      pos3 = e.clientX;
      pos4 = e.clientY;
      panel.style.top = panel.offsetTop - pos2 + "px";
      panel.style.left = panel.offsetLeft - pos1 + "px";
    };
    function closeDragElement() {
      document.removeEventListener("mouseup", closeDragElement);
      document.removeEventListener("mousemove", elementDrag);
      sessionStorage.setItem(
        "@bmsmnt/scrollytelling-visualizer:position",
        JSON.stringify({
          top: panel == null ? void 0 : panel.style.top,
          left: panel == null ? void 0 : panel.style.left
        })
      );
    }
    panelHeader.addEventListener("mousedown", dragMouseDown);
    setInitialized(true);
    return () => {
      panelHeader.removeEventListener("mousedown", dragMouseDown);
      closeDragElement();
    };
  }, []);
  useEffect(() => {
    return internalEventEmmiter.on("timeline:refresh", () => {
      const roots2 = [];
      gsap.globalTimeline.getChildren().forEach((t) => {
        var _a2;
        if (!((_a2 = t.data) == null ? void 0 : _a2.isScrollytellingTween))
          return;
        const data = t.data;
        switch (data.type) {
          case "root": {
            const existingRootItem = roots2.find((r) => r.id === data.id);
            if (!existingRootItem) {
              roots2.push({
                id: data.id,
                debug: data.debug,
                label: data.label,
                children: [],
                tween: t
              });
            } else {
              existingRootItem.debug = data.debug;
              existingRootItem.tween = t;
              existingRootItem.label = data.label;
            }
            break;
          }
          case "rest":
          case "waypoint":
          case "animation": {
            const existingRootItem = roots2.find((r) => r.id === data.rootId);
            if (!existingRootItem) {
              roots2.push({
                id: data.rootId,
                debug: false,
                label: data.rootId,
                children: [t],
                tween: void 0
              });
            } else {
              existingRootItem.children.push(t);
            }
            break;
          }
        }
      });
      roots2.sort((a, b) => {
        var _a2, _b, _c, _d;
        const aTriggerEl = (_b = (_a2 = a.tween) == null ? void 0 : _a2.scrollTrigger) == null ? void 0 : _b.trigger;
        const bTriggerEl = (_d = (_c = b.tween) == null ? void 0 : _c.scrollTrigger) == null ? void 0 : _d.trigger;
        if (!aTriggerEl || !bTriggerEl)
          return 0;
        const aTop = aTriggerEl.getBoundingClientRect().top;
        const bTop = bTriggerEl.getBoundingClientRect().top;
        return aTop - bTop;
      });
      setRoots(roots2.filter((r) => r.debug));
    });
  }, []);
  useEffect(() => {
    internalEventEmmiter.emit("timeline:refresh");
  }, []);
  if (dismiss)
    return /* @__PURE__ */ jsx(Fragment, {});
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: clsx(visualizer_module_default["root"], initialized && visualizer_module_default["initialized"]),
      ref: panelRef,
      children: [
        /* @__PURE__ */ jsxs("header", { className: visualizer_module_default["header"], ref: panelHeaderRef, children: [
          /* @__PURE__ */ jsxs("div", { className: visualizer_module_default["actions"], children: [
            /* @__PURE__ */ jsx(
              Select,
              {
                value: selectedRoot,
                onChange: (e) => setSelectedRoot(e.currentTarget.value),
                onPointerDown: (e) => e.stopPropagation(),
                children: roots.map((r) => {
                  return /* @__PURE__ */ jsx("option", { value: r.id, children: r.label }, r.id);
                })
              }
            ),
            /* @__PURE__ */ jsx("div", { className: visualizer_module_default["actions"], children: /* @__PURE__ */ jsxs(
              "button",
              {
                className: clsx(visualizer_module_default["button"], visualizer_module_default["scrollToRoot"]),
                onClick: () => {
                  var _a2, _b;
                  const triggerElement = (_b = (_a2 = root == null ? void 0 : root.tween) == null ? void 0 : _a2.scrollTrigger) == null ? void 0 : _b.trigger;
                  if (triggerElement) {
                    setIsUserScroll(false);
                    setTimeout(
                      () => {
                        setIsUserScroll(true);
                      },
                      1500
                    );
                    triggerElement.scrollIntoView({
                      behavior: "smooth"
                    });
                  }
                },
                children: [
                  /* @__PURE__ */ jsx(
                    "svg",
                    {
                      width: "12",
                      height: "12",
                      viewBox: "0 0 12 12",
                      fill: "none",
                      xmlns: "http://www.w3.org/2000/svg",
                      children: /* @__PURE__ */ jsx(
                        "path",
                        {
                          d: "M10.875 6.00004L6.398 1.52254C6.178 1.30304 5.822 1.30304 5.6025 1.52254L1.125 6.00004M9.75 4.87504V9.93754C9.75 10.248 9.498 10.5 9.1875 10.5H7.125V8.06254C7.125 7.75204 6.873 7.50004 6.5625 7.50004H5.4375C5.127 7.50004 4.875 7.75204 4.875 8.06254V10.5H2.8125C2.502 10.5 2.25 10.248 2.25 9.93754V4.87504M7.875 10.5H3.75",
                          stroke: "white",
                          strokeLinecap: "round",
                          strokeLinejoin: "round"
                        }
                      )
                    }
                  ),
                  /* @__PURE__ */ jsx("span", { className: visualizer_module_default["text"], children: "SCROLL TO ROOT" })
                ]
              }
            ) })
          ] }),
          /* @__PURE__ */ jsxs("div", { className: visualizer_module_default["actions"], children: [
            /* @__PURE__ */ jsx(
              "button",
              {
                className: visualizer_module_default["button"],
                onClick: () => setMinimize((p) => !p),
                children: minimize ? /* @__PURE__ */ jsxs(
                  "svg",
                  {
                    viewBox: "0 0 16 16",
                    fill: "none",
                    xmlns: "http://www.w3.org/2000/svg",
                    children: [
                      /* @__PURE__ */ jsx(
                        "path",
                        {
                          fillRule: "evenodd",
                          clipRule: "evenodd",
                          d: "M2.43359 8C2.43359 7.72386 2.65745 7.5 2.93359 7.5H13.0669C13.3431 7.5 13.5669 7.72386 13.5669 8C13.5669 8.27614 13.3431 8.5 13.0669 8.5H2.93359C2.65745 8.5 2.43359 8.27614 2.43359 8Z",
                          fill: "white"
                        }
                      ),
                      /* @__PURE__ */ jsx(
                        "path",
                        {
                          fillRule: "evenodd",
                          clipRule: "evenodd",
                          d: "M8 2.43311C8.27614 2.43311 8.5 2.65697 8.5 2.93311L8.5 13.0664C8.5 13.3426 8.27614 13.5664 8 13.5664C7.72386 13.5664 7.5 13.3426 7.5 13.0664L7.5 2.93311C7.5 2.65697 7.72386 2.43311 8 2.43311Z",
                          fill: "white"
                        }
                      )
                    ]
                  }
                ) : /* @__PURE__ */ jsx(
                  "svg",
                  {
                    viewBox: "0 0 16 16",
                    fill: "none",
                    xmlns: "http://www.w3.org/2000/svg",
                    children: /* @__PURE__ */ jsx(
                      "path",
                      {
                        fillRule: "evenodd",
                        clipRule: "evenodd",
                        d: "M2.43359 8C2.43359 7.72386 2.65745 7.5 2.93359 7.5H13.0669C13.3431 7.5 13.5669 7.72386 13.5669 8C13.5669 8.27614 13.3431 8.5 13.0669 8.5H2.93359C2.65745 8.5 2.43359 8.27614 2.43359 8Z",
                        fill: "white"
                      }
                    )
                  }
                )
              }
            ),
            /* @__PURE__ */ jsx("button", { className: visualizer_module_default["button"], onClick: () => setDismiss(true), children: /* @__PURE__ */ jsx(
              "svg",
              {
                viewBox: "0 0 16 16",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
                children: /* @__PURE__ */ jsx(
                  "path",
                  {
                    fillRule: "evenodd",
                    clipRule: "evenodd",
                    d: "M12.5402 4.27333C12.7648 4.04878 12.7648 3.68471 12.5402 3.46016C12.3157 3.23561 11.9516 3.23561 11.7271 3.46016L8.00033 7.18691L4.27358 3.46016C4.04903 3.23561 3.68496 3.23561 3.46041 3.46016C3.23585 3.68471 3.23585 4.04878 3.46041 4.27333L7.18715 8.00008L3.46041 11.7268C3.23585 11.9514 3.23585 12.3154 3.46041 12.54C3.68496 12.7646 4.04903 12.7646 4.27358 12.54L8.00033 8.81325L11.7271 12.54C11.9516 12.7646 12.3157 12.7646 12.5402 12.54C12.7648 12.3154 12.7648 11.9514 12.5402 11.7268L8.8135 8.00008L12.5402 4.27333Z",
                    fill: "white"
                  }
                )
              }
            ) })
          ] })
        ] }),
        !minimize && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx("main", { className: visualizer_module_default["main"], children: /* @__PURE__ */ jsxs("div", { className: visualizer_module_default["timeline"], children: [
            /* @__PURE__ */ jsxs("div", { className: visualizer_module_default["guides"], children: [
              /* @__PURE__ */ jsx("div", { className: visualizer_module_default["guide"], style: { left: "0%" }, children: /* @__PURE__ */ jsx("span", { className: visualizer_module_default["percent"], children: "0%" }) }),
              /* @__PURE__ */ jsx("div", { className: visualizer_module_default["guide"], style: { left: "25%" }, children: /* @__PURE__ */ jsx("span", { className: visualizer_module_default["percent"], children: "25%" }) }),
              /* @__PURE__ */ jsx("div", { className: visualizer_module_default["guide"], style: { left: "50%" }, children: /* @__PURE__ */ jsx("span", { className: visualizer_module_default["percent"], children: "50%" }) }),
              /* @__PURE__ */ jsx("div", { className: visualizer_module_default["guide"], style: { left: "75%" }, children: /* @__PURE__ */ jsx("span", { className: visualizer_module_default["percent"], children: "75%" }) }),
              /* @__PURE__ */ jsx("div", { className: visualizer_module_default["guide"], style: { left: "100%" }, children: /* @__PURE__ */ jsx("span", { className: visualizer_module_default["percent"], children: "100%" }) })
            ] }),
            /* @__PURE__ */ jsx("div", { className: visualizer_module_default["tweens"], children: root == null ? void 0 : root.children.map((t, idx) => {
              const data = t.data;
              if (data.type === "animation") {
                return /* @__PURE__ */ jsx("div", { className: visualizer_module_default["row"], children: /* @__PURE__ */ jsx(Tween, { tween: t, root, idx }) }, idx);
              }
              if (data.type === "waypoint") {
                return /* @__PURE__ */ jsx("div", { className: visualizer_module_default["row"], children: /* @__PURE__ */ jsx(Waypoint, { tween: t, root, idx }) }, idx);
              }
              return /* @__PURE__ */ jsx(Fragment, {});
            }) }),
            /* @__PURE__ */ jsxs("div", { className: visualizer_module_default["progress"], children: [
              /* @__PURE__ */ jsx("div", { className: visualizer_module_default["marker"], ref: markerRef, children: /* @__PURE__ */ jsxs("span", { className: visualizer_module_default["thumb"], children: [
                /* @__PURE__ */ jsx("span", { className: visualizer_module_default["percent"], children: /* @__PURE__ */ jsx(ProgressStatus, { root }) }),
                /* @__PURE__ */ jsxs(
                  "svg",
                  {
                    width: "8",
                    height: "11",
                    viewBox: "0 0 8 11",
                    fill: "none",
                    xmlns: "http://www.w3.org/2000/svg",
                    children: [
                      /* @__PURE__ */ jsx("mask", { id: "path-1-inside-1_2793_1632", fill: "white", children: /* @__PURE__ */ jsx(
                        "path",
                        {
                          fillRule: "evenodd",
                          clipRule: "evenodd",
                          d: "M8 0H0V8L4 11L8 8V0Z"
                        }
                      ) }),
                      /* @__PURE__ */ jsx(
                        "path",
                        {
                          fillRule: "evenodd",
                          clipRule: "evenodd",
                          d: "M8 0H0V8L4 11L8 8V0Z",
                          fill: "white"
                        }
                      ),
                      /* @__PURE__ */ jsx(
                        "path",
                        {
                          d: "M0 0V-1H-1V0H0ZM8 0H9V-1H8V0ZM0 8H-1V8.5L-0.6 8.8L0 8ZM4 11L3.4 11.8L4 12.25L4.6 11.8L4 11ZM8 8L8.6 8.8L9 8.5V8H8ZM0 1H8V-1H0V1ZM1 8V0H-1V8H1ZM4.6 10.2L0.6 7.2L-0.6 8.8L3.4 11.8L4.6 10.2ZM7.4 7.2L3.4 10.2L4.6 11.8L8.6 8.8L7.4 7.2ZM7 0V8H9V0H7Z",
                          fill: "white",
                          mask: "url(#path-1-inside-1_2793_1632)"
                        }
                      )
                    ]
                  }
                )
              ] }) }),
              /* @__PURE__ */ jsx("div", { className: visualizer_module_default["trail"], children: /* @__PURE__ */ jsx("div", { className: visualizer_module_default["gradient"], ref: trailRef }) })
            ] })
          ] }) }),
          /* @__PURE__ */ jsxs("footer", { className: visualizer_module_default["footer"], children: [
            /* @__PURE__ */ jsxs("span", { children: [
              "Visualizer",
              /* @__PURE__ */ jsxs("span", { className: visualizer_module_default["version"], children: [
                " (v.",
                package_default.version,
                ")"
              ] })
            ] }),
            /* @__PURE__ */ jsxs("span", { children: [
              "made with \u{1F5A4} by",
              " ",
              /* @__PURE__ */ jsx(
                "a",
                {
                  href: "https://basement.studio",
                  target: "_blank",
                  className: visualizer_module_default["bsmnt"],
                  rel: "noopener",
                  children: "bsmnt"
                }
              ),
              "."
            ] })
          ] })
        ] })
      ]
    }
  );
};
function Debugger() {
  const [mountInstance, setMountInstance] = useState(false);
  useEffect(() => {
    const w = window;
    const alreadyMountedInstance = w.__scrollytelling_alreadyMountedDebuggerInstance;
    if (alreadyMountedInstance)
      return;
    setMountInstance(true);
    w.__scrollytelling_alreadyMountedDebuggerInstance = true;
  }, []);
  if (!mountInstance)
    return /* @__PURE__ */ jsx(Fragment, {});
  return /* @__PURE__ */ jsx(Portal, { children: /* @__PURE__ */ jsx(Visualizer, {}) });
}

export { Debugger as default };
